File: src/main/java/com/github/kornilova_l/flamegraph/plugin/server/trees/ser_trees/call_tree/CallTreesBuilder.java
package com.github.kornilova_l.flamegraph.plugin.server.trees.ser_trees.call_tree;

import com.github.kornilova_l.flamegraph.proto.EventProtos;
import com.github.kornilova_l.flamegraph.proto.TreeProtos;
import com.github.kornilova_l.flamegraph.proto.TreesProtos;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CallTreesBuilder {
    private static final com.intellij.openapi.diagnostic.Logger LOG =
            com.intellij.openapi.diagnostic.Logger.getInstance(CallTreesBuilder.class);
    private Map<Long, CTBuilder> treesMap = new HashMap<>();
    private Map<Long, String> classNames = new HashMap<>();
    private Map<Long, String> threadsNames = new HashMap<>();
    @Nullable
    private TreesProtos.Trees trees = null;

    public CallTreesBuilder(File logFile) {
        try (InputStream inputStream = new FileInputStream(logFile)) {
            processEvents(inputStream);
            long startTimeOfFirstThread = getStartTimeOfFirstThread(treesMap);
            trees = HashMapToTrees(treesMap, startTimeOfFirstThread);
        } catch (IOException e) {
            LOG.error(e);
        }
    }

    @Nullable
    private static TreesProtos.Trees HashMapToTrees(Map<Long, CTBuilder> trees,
                                                              long startTimeOfFirstThread) {
        TreesProtos.Trees.Builder treesBuilder = TreesProtos.Trees.newBuilder();
        for (CTBuilder oTBuilder : trees.values()) {
            TreeProtos.Tree tree = oTBuilder.getBuiltTree(startTimeOfFirstThread);
            if (tree != null) {
                treesBuilder.addTrees(
                        tree
                );
            }
        }
        if (treesBuilder.getTreesCount() == 0) {
            return null;
        }
        return treesBuilder.build();
    }

    private void processEvents(InputStream inputStream) throws IOException {
        EventProtos.Event event = EventProtos.Event.parseDelimitedFrom(inputStream);
        while (event != null) {
            switch (event.getTypeCase()) {
                case METHODEVENT:
                    addMethodEvent(event);
                    break;
                case NEWCLASS:
                    registerClass(event);
                    break;
                case NEWTHREAD:
                    threadsNames.put(event.getNewThread().getId(), event.getNewThread().getName());
                    break;
                case TYPE_NOT_SET:
                default:
                    throw new RuntimeException("Event without type");
            }
            event = EventProtos.Event.parseDelimitedFrom(inputStream);
        }
    }

    private void registerClass(EventProtos.Event event) {
        String className = event.getNewClass().getName();
        classNames.put(
                event.getNewClass().getId(),
                className.replace('/', '.'));
    }

    private void addMethodEvent(EventProtos.Event event) {
        EventProtos.Event.MethodEvent methodEvent = event.getMethodEvent();
        CTBuilder ctBuilder = getCTBuilder(methodEvent);
        String className = classNames.get(methodEvent.getClassNameId());
        if (className == null) {
            throw new RuntimeException("Class name is not known");
        }
        ctBuilder.addEvent(methodEvent, className);
    }

    private CTBuilder getCTBuilder(EventProtos.Event.MethodEvent methodEvent) {
        String threadName = threadsNames.get(methodEvent.getThreadId());
        if (threadName == null) {
            LOG.debug("Thread name is not known. MethodEvent: " + methodEvent.toString());
            return treesMap.computeIfAbsent(
                    methodEvent.getThreadId(),
                    k -> new CTBuilder(
                            methodEvent.getStartTime(),
                            ""
                    )
            );
        }
        return treesMap.computeIfAbsent(
                methodEvent.getThreadId(),
                k -> new CTBuilder(
                        methodEvent.getStartTime(),
                        threadName
                )
        );
    }

    private long getStartTimeOfFirstThread(Map<Long, CTBuilder> treesMap) {
        List<CTBuilder> treesList = new ArrayList<>(treesMap.values());
        int size = treesList.size();
        if (size == 0) {
            return 0;
        }
        long startTimeOfFirstThread = treesList.get(0).getThreadStartTime();
        for (int i = 1; i < size; i++) {
            if (treesList.get(i).getThreadStartTime() < startTimeOfFirstThread) {
                startTimeOfFirstThread = treesList.get(i).getThreadStartTime();
            }
        }
        return startTimeOfFirstThread;
    }

    @Nullable
    public TreesProtos.Trees getTrees() {
        return trees;
    }
}

File: src/main/java/com/github/kornilova_l/flamegraph/plugin/server/converters/calltree/fierix/CallTreesBuilder.kt
package com.github.kornilova_l.flamegraph.plugin.server.converters.calltree.fierix

import com.github.kornilova_l.flamegraph.plugin.pleaseReportIssue
import com.github.kornilova_l.flamegraph.proto.EventProtos
import com.github.kornilova_l.flamegraph.proto.TreesProtos
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.*

class CallTreesBuilder(logFile: File) {
    private val treesMap = HashMap<Long, CTBuilder>()
    private val classNames = HashMap<Long, String>()
    private val threadsNames = HashMap<Long, String>()
    var trees: TreesProtos.Trees? = null
        private set

    init {
        try {
            FileInputStream(logFile).use { inputStream ->
                processEvents(inputStream)
                val startTimeOfFirstThread = getStartTimeOfFirstThread(treesMap)
                trees = hashMapToTrees(treesMap, startTimeOfFirstThread)
            }
        } catch (e: IOException) {
            LOG.error(e)
        }

    }

    @Throws(IOException::class)
    private fun processEvents(inputStream: InputStream) {
        var event: EventProtos.Event? = EventProtos.Event.parseDelimitedFrom(inputStream)
        while (event != null) {
            when (event.typeCase) {
                EventProtos.Event.TypeCase.METHODEVENT -> addMethodEvent(event)
                EventProtos.Event.TypeCase.NEWCLASS -> registerClass(event)
                EventProtos.Event.TypeCase.NEWTHREAD -> threadsNames[event.newThread.id] = event.newThread.name
                EventProtos.Event.TypeCase.TYPE_NOT_SET -> throw RuntimeException("$pleaseReportIssue: Event without type")
                else -> throw RuntimeException("$pleaseReportIssue: Event without type")
            }
            event = EventProtos.Event.parseDelimitedFrom(inputStream)
        }
    }

    private fun registerClass(event: EventProtos.Event) {
        val className = event.newClass.name
        classNames[event.newClass.id] = className.replace('/', '.')
    }

    private fun addMethodEvent(event: EventProtos.Event) {
        val methodEvent = event.methodEvent
        val ctBuilder = getCTBuilder(methodEvent)
        val className = classNames[methodEvent.classNameId]
                ?: throw RuntimeException("$pleaseReportIssue (also please upload this ser file): Class name is not known. id = ${methodEvent.classNameId}")
        ctBuilder.addEvent(methodEvent, className)
    }

    private fun getCTBuilder(methodEvent: EventProtos.Event.MethodEvent): CTBuilder {
        val threadName = threadsNames[methodEvent.threadId]
        if (threadName == null) {
            LOG.debug("Thread name is not known. MethodEvent: " + methodEvent.toString())
            return treesMap.computeIfAbsent(
                    methodEvent.threadId
            ) { CTBuilder(methodEvent.startTime, "") }
        }
        return treesMap.computeIfAbsent(
                methodEvent.threadId
        ) { CTBuilder(methodEvent.startTime, threadName) }
    }

    private fun getStartTimeOfFirstThread(treesMap: Map<Long, CTBuilder>): Long {
        val treesList = ArrayList(treesMap.values)
        val size = treesList.size
        if (size == 0) {
            return 0
        }
        var startTimeOfFirstThread = treesList[0].threadStartTime
        for (i in 1 until size) {
            if (treesList[i].threadStartTime < startTimeOfFirstThread) {
                startTimeOfFirstThread = treesList[i].threadStartTime
            }
        }
        return startTimeOfFirstThread
    }

    companion object {
        private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(CallTreesBuilder::class.java)

        private fun hashMapToTrees(trees: Map<Long, CTBuilder>,
                                   startTimeOfFirstThread: Long): TreesProtos.Trees? {
            val treesBuilder = TreesProtos.Trees.newBuilder()
            for (oTBuilder in trees.values) {
                val tree = oTBuilder.getBuiltTree(startTimeOfFirstThread)
                if (tree != null) {
                    treesBuilder.addTrees(
                            tree
                    )
                }
            }
            return if (treesBuilder.treesCount == 0) {
                null
            } else treesBuilder.build()
        }
    }
}
